above_ground.py

```diff
@@ -138,6 +169,238 @@ def stringify(v):
+
+        return n
+
+    def safeCreateGetEdge(sourceKey, name, fromNodeId, toNodeId, tags=None):
+        try:
+            n = xp_state.gc.get_edge(sourceKey)
+            if n is None:
+                n = xp_state.gc.create_edge(sourceKey, name, fromNodeId, toNodeId, tags)
+        except:
+            n = xp_state.gc.create_edge(sourceKey, name, fromNodeId, toNodeId, tags)
+
+        return n
+
+    def safeCreateGetNodeVersion(sourceKey):
+        # Good for singleton node versions
+        try:
+            n = xp_state.gc.get_node_latest_versions(sourceKey)
+            if n is None or n == []:
+                n = xp_state.gc.create_node_version(xp_state.gc.get_node(sourceKey).get_id())
+            else:
+                assert len(n) == 1
+                return xp_state.gc.get_node_version(n[0])
+        except:
+            n = xp_state.gc.create_node_version(xp_state.gc.get_node(sourceKey).get_id())
+
+        return n
+
+    def safeCreateLineage(sourceKey, name, tags=None):
+        try:
+            n = xp_state.gc.create_lineage_edge(sourceKey, name, tags)
+        except:
+            n = xp_state.gc.create_lineage_edge(sourceKey, name, tags)
+
+        return n
+
+    def stringify(v):
+         # https://stackoverflow.com/a/22505259/9420936
+        return hashlib.md5(json.dumps(str(v) , sort_keys=True).encode('utf-8')).hexdigest()
+
+    def get_sha(directory):
+        output = subprocess.check_output('git log -1 --format=format:%H'.split()).decode()
+
+
+    def geteg(xp_state, inputCH):
+        original = os.getcwd()
+        os.chdir(xp_state.versioningDirectory + '/' + xp_state.EXPERIMENT_NAME)
+        util.runProc('git checkout ' + inputCH)
+        os.chdir("0/")
+        with open('experiment_graph.pkl', 'rb') as f:
+            experimentg = dill.load(f)
+        # for each in experimentg.d.keys():
+        #     temp = experimentg.d[each]
+        #     for x in temp:
+        #         if type(x) is not set and type(x) is not Action:
+        #             print(x.getLocation())
+        #             input()
+        util.runProc('git checkout master')
+        os.chdir(original)
+        return experimentg
+
+    sourcekeySpec = 'flor.' + xp_state.EXPERIMENT_NAME
+    specnode = safeCreateGetNode(sourcekeySpec, "null")
+
+    #gives you a list of most recent experiment versions
+    latest_experiment_node_versions = xp_state.gc.get_node_latest_versions(sourcekeySpec)
+    if latest_experiment_node_versions == []:
+        latest_experiment_node_versions = None
+
+    # print(sourcekeySpec)
+    # print(xp_state.gc.get_node_history(sourcekeySpec))
+    # print(xp_state.gc.get_node_latest_versions('flor.plate_demo'))
+    # input()
+    timestamps = [xp_state.gc.get_node_version(x).get_tags()['timestamp']['value'] for x in latest_experiment_node_versions]
+    latest_experiment_nodev = latest_experiment_node_versions[timestamps.index(min(timestamps))]
```

Comment:
Why is this correct? Why do we care about the latest (in time)? Also, by chaining commits, we should have it so that len(latest_experiment_node_versions) is 1. We should discuss in meeting.